cmd/sync/azure.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/compute/mgmt/compute"
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/network/mgmt/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// AzureClient allows you to get the list of IP addresses of VirtualMachines of a VirtualMachine Scale Set. It implements the CloudProvider interface
+type AzureClient struct {
+	config      *azureConfig
+	vmssClient  compute.VirtualMachineScaleSetsClient
+	ifaceClient network.InterfacesClient
+	authorizer  autorest.Authorizer
+}
+
+// NewAzureClient creates an AzureClient
+func NewAzureClient(data []byte) (*AzureClient, error) {
+	azureClient := &AzureClient{}
+	cfg, err := azureClient.parseAzureConfig(data)
+	if err != nil {
+		return nil, fmt.Errorf("error validating config: %v", err)
+	}
+
+	azureClient.config = cfg
+
+	err = azureClient.configure()
+	if err != nil {
+		return nil, fmt.Errorf("error configuring Azure Client: %v", err)
+	}
+
+	return azureClient, nil
+}
+
+// parseAzureConfig parses and validates AzureClient config
+func (client *AzureClient) parseAzureConfig(data []byte) (*azureConfig, error) {
+	cfg := &azureConfig{}
+	err := yaml.Unmarshal(data, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	err = validateAzureConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// GetPrivateIPsForScalingGroup returns the list of IP addresses of instances of the Virtual Machine Scale Set
+func (client *AzureClient) GetPrivateIPsForScalingGroup(name string) ([]string, error) {
+	var ips []string
+
+	ctx := context.Background()
+
+	for iFaces, err := client.ifaceClient.ListVirtualMachineScaleSetNetworkInterfaces(ctx, client.config.ResourceGroupName, name); iFaces.NotDone() || err != nil; err = iFaces.NextWithContext(ctx) {
+		if err != nil {
+			return nil, err
+		}
+
+		for _, iFace := range iFaces.Values() {
+			if iFace.VirtualMachine != nil && iFace.VirtualMachine.ID != nil {
+				for _, n := range *iFace.IPConfigurations {
+					if *n.Primary && n.InterfaceIPConfigurationPropertiesFormat.PrivateIPAddress != nil {
+						ips = append(ips, *n.InterfaceIPConfigurationPropertiesFormat.PrivateIPAddress)
+						break
+					}
+				}
+			}
+		}
+	}
+	return ips, nil
+}
+
+// CheckIfScalingGroupExists checks if the Virtual Machine Scale Set exists
+func (client *AzureClient) CheckIfScalingGroupExists(name string) (bool, error) {
+	ctx := context.Background()
+	vmss, err := client.vmssClient.Get(ctx, client.config.ResourceGroupName, name)
+	if err != nil {
+		return false, err
```

Comment:
I think it depends, this function is called form `main.go` where we print something like `Couldn't check if scaling group existed: <err>` I think that's enough information. Adding something like `there was a problem creating the VMSS client` adds 0 information to be honest, and makes logs more difficult to read. 

Most of the errors here are caused by wrong auth, wrong resource group name, problem in connection, etc. If you think this is a must and really important I can add something

Reply:
I'll add a text

I think in line 100 is different cos `configure` is called from another function where we handle the new error. It seems weird that we need to add information about all errors. In the end it will create a lot of noise in the logs. Let me know if that makes sense.